File: src/geocode.ts
/**
 * Reverse geocoding via Nominatim (OpenStreetMap).
 *
 * Their usage policy allows this only under strict conditions: at most one
 * request per second, results must be cached, and the client has to identify
 * itself. A browser cannot set `User-Agent` (the header is forbidden by fetch),
 * so identification falls back to the Referer the browser sends automatically,
 * optionally strengthened by the `email` parameter Nominatim documents for
 * exactly this case.
 */

const ENDPOINT = "https://nominatim.openstreetmap.org/reverse";
const MIN_REQUEST_GAP_MS = 1100;
/*
 * The `:2` is a version, and bumping it throws the stored labels away.
 *
 * Needed the moment the integration learned to name a shopping centre instead
 * of the street around it: this cache is consulted before anything is asked,
 * so a stay resolved last week would go on reading "Grinzinger Straße 112"
 * for a month no matter what the integration now answers.
 *
 * Raised to 3 for a second helping of the same problem. The integration takes
 * care not to store an address it only fell back to while the lookup service
 * was busy -- and this cache stored it anyway, which is how a stay at the
 * Bauhaus went on reading "Jägerstraße 82" after the fix was already in.
 */
const CACHE_KEY = "family-tracking-card:geocode:3";
const CACHE_TTL_MS = 30 * 24 * 60 * 60 * 1000;
/** ~11 m at the equator: fine enough for a stay, coarse enough to reuse. */
const CACHE_PRECISION = 4;

export interface GeocodeOptions {
  /** Contact address forwarded to Nominatim, see their usage policy. */
  email?: string;
  language?: string;
  /**
   * Whether to ask what the coordinate lies inside. With it, a stay in a
   * shopping centre reads as the centre rather than as the street beside it or
   * whichever shop happens to be nearest. Off, nothing is asked of the second
   * service at all.
   */
  places?: boolean;
  /**
   * Whether to keep the address as well, in brackets after the name. Part of
   * the cache key for the same reason `places` is: the same coordinate has
   * more than one right answer, and a stored one from the other setting would
   * make the switch look broken.
   */
  placeAddress?: boolean;
  /**
   * Home Assistant's websocket call. When the Family Tracking integration is
   * installed it answers `family_tracking/geocode`, and then the lookup belongs
   * there: one cache for every browser in the house instead of one per device,
   * one queue honouring the rate limit, and results that survive a cleared
   * browser storage. Without the integration the card asks Nominatim itself,
   * exactly as before.
   */
  callWS?: <T>(message: Record<string, unknown>) => Promise<T>;
}

/** The instance-wide settings of the integration that the card follows. */
export interface ServerSettings {
  /** `false` when lookups are switched off for the whole instance. */
  geocode?: boolean;
  /** Fixes reporting a worse accuracy than this, in metres, are left out. */
  maxAccuracy?: number;
}

/**
 * Asks the integration for its settings.
 *
 * Every field is `undefined` when there is no integration to ask or it could
 * not answer: the card then decides on its own, as it always did.
 */
export async function fetchServerSettings(
  callWS: GeocodeOptions["callWS"]
): Promise<ServerSettings> {
  if (!callWS) return {};
  try {
    const result = await callWS<{ geocode?: unknown; max_accuracy?: unknown } | null>({
      type: "family_tracking/settings",
    });
    const maxAccuracy = Number(result?.max_accuracy);
    return {
      geocode: typeof result?.geocode === "boolean" ? result.geocode : undefined,
      maxAccuracy: Number.isFinite(maxAccuracy) && maxAccuracy > 0 ? maxAccuracy : undefined,
    };
  } catch {
    return {};
  }
}

/**
 * Whether the integration lets this instance resolve addresses at all.
 *
 * Only an explicit `false` means the editor's lookup switches have no effect.
 */
export async function serverGeocodeEnabled(
  callWS: GeocodeOptions["callWS"]
): Promise<boolean | undefined> {
  return (await fetchServerSettings(callWS)).geocode;
}

/** Remembers a missing integration, so the card asks once and not per stay. */
let serverGeocoding: boolean | undefined;

/**
 * Remembers that the instance switched lookups off.
 *
 * Kept apart from `serverGeocoding`, because the two lead opposite ways: a
 * missing integration means fall back to Nominatim, a switched-off one means
 * do not.
 */
let serverDisabled = false;

/** Test helper: forget whether the integration answered. */
export function resetServerGeocoding(): void {
  serverGeocoding = undefined;
  serverDisabled = false;
}

/** What the integration answered: the label, and whether it is the final one. */
interface ServerAnswer {
  label?: string;
  settled: boolean;
  /** The instance resolves no addresses at all; do not go around it. */
  disabled?: boolean;
}

async function viaServer(
  lat: number,
  lon: number,
  options: GeocodeOptions
): Promise<ServerAnswer | null> {
  if (!options.callWS || serverGeocoding === false) return null;
  if (serverDisabled) return { settled: true, disabled: true };
  try {
    const result = await options.callWS<{
      label?: string;
      settled?: boolean;
      disabled?: boolean;
    } | null>({
      type: "family_tracking/geocode",
      latitude: lat,
      longitude: lon,
      language: options.language,
      places: options.places !== false,
      address: options.placeAddress === true,
    });
    serverGeocoding = true;
    if (result?.disabled) {
      serverDisabled = true;
      return { settled: true, disabled: true };
    }
    // `settled: false` means the integration wanted to name the place and
    // could not ask -- the label is an address standing in for a name that is
    // not available this minute.
    return { label: result?.label || undefined, settled: result?.settled !== false };
  } catch (err) {
    // Not the same thing twice. `unknown_command` means the integration is not
    // installed, and then asking again for every stay only costs a round trip
    // each time. Anything else -- most often "not_ready" while Home Assistant
    // is still starting up -- is over in a minute, and giving up on it for the
    // rest of the page would hand back the worse answer all evening.
    if ((err as { code?: string } | undefined)?.code === "unknown_command") {
      serverGeocoding = false;
    }
    return null;
  }
}

interface CacheEntry {
  label: string;
  at: number;
  /**
   * Set when the label came from asking Nominatim directly rather than from
   * the integration. Those answers know nothing about shopping centres, so
   * they are kept for this page and no longer: written to storage, a lookup
   * made while Home Assistant was still starting would outlive the reason it
   * was ever needed.
   */
  direct?: boolean;
}

export const cacheKeyFor = (lat: number, lon: number): string =>
  `${lat.toFixed(CACHE_PRECISION)},${lon.toFixed(CACHE_PRECISION)}`;

/**
 * The cache key for one lookup.
 *
 * The setting is part of it. The same coordinate has two right answers -- the
 * place and the address -- and without this, turning the option off would go
 * on showing the place it had already stored, which reads like the switch does
 * nothing.
 */
const entryKeyFor = (lat: number, lon: number, places: boolean, withAddress: boolean): string => {
  const base = cacheKeyFor(lat, lon);
  if (!places) return `${base}|addr`;
  return withAddress ? `${base}|both` : base;
};

/**
 * Where the house number goes before the street name rather than after it.
 *
 * "350 5th Avenue" in New York, "Pariser Platz 1" in Berlin -- and Nominatim
 * hands both over as separate fields, so the order is ours to get right.
 * Countries not listed take the number after the street, which covers most of
 * Europe and South America. Kept in step with `HOUSE_NUMBER_FIRST` in
 * `custom_components/family_tracking/address.py`.
 */
const HOUSE_NUMBER_FIRST = new Set([
  "us", "ca", "gb", "ie", "au", "nz", "fr", "in", "sg", "my", "hk", "ph", "th", "za",
]);

/** Street and house number in the order that country writes them. */
export function streetHead(street: string, houseNumber?: string, countryCode?: string): string {
  if (!street) return "";
  if (!houseNumber) return street;
  return HOUSE_NUMBER_FIRST.has((countryCode ?? "").toLowerCase())
    ? `${houseNumber} ${street}`
    : `${street} ${houseNumber}`;
}

/**
 * Condenses a Nominatim address object into one readable line.
 * Pure function, kept separate so it can be tested without network access.
 */
export function shortLabel(result: any): string | undefined {
  const address = result?.address ?? {};
  const street = address.road ?? address.pedestrian ?? address.footway ?? address.path;
  const place =
    address.village ?? address.town ?? address.city ?? address.municipality ?? address.county;

  if (street) {
    const head = streetHead(street, address.house_number, address.country_code);
    return place ? `${head}, ${place}` : head;
  }
  const named = result?.name || address.amenity || address.shop || address.building;
  if (named) return place ? `${named}, ${place}` : named;
  if (place) return place;
  return typeof result?.display_name === "string"
    ? result.display_name.split(",").slice(0, 2).join(",").trim()
    : undefined;
}

let memoryCache: Map<string, CacheEntry> | undefined;

function cache(): Map<string, CacheEntry> {
  if (memoryCache) return memoryCache;
  memoryCache = new Map();
  try {
    const raw = window.localStorage.getItem(CACHE_KEY);
    if (raw) {
      const now = Date.now();
      for (const [key, entry] of Object.entries(JSON.parse(raw) as Record<string, CacheEntry>)) {
        if (entry && now - entry.at < CACHE_TTL_MS) memoryCache.set(key, entry);
      }
    }
  } catch {
    // A corrupt or unavailable localStorage must not break the card.
  }
  return memoryCache;
}

let persistHandle: number | undefined;

function persist(): void {
  if (persistHandle !== undefined) return;
  persistHandle = window.setTimeout(() => {
    persistHandle = undefined;
    try {
      const lasting = [...cache()].filter(([, entry]) => !entry.direct);
      window.localStorage.setItem(CACHE_KEY, JSON.stringify(Object.fromEntries(lasting)));
    } catch {
      // Quota exceeded or private mode: the in-memory cache still works.
    }
  }, 1000);
}

/** Serialises all lookups and keeps at least MIN_REQUEST_GAP_MS between them. */
let queue: Promise<unknown> = Promise.resolve();
let lastRequestAt = 0;

function schedule<T>(task: () => Promise<T>): Promise<T> {
  const run = queue.then(async () => {
    const wait = lastRequestAt + MIN_REQUEST_GAP_MS - Date.now();
    if (wait > 0) await new Promise((resolve) => setTimeout(resolve, wait));
    lastRequestAt = Date.now();
    return task();
  });
  queue = run.catch(() => undefined);
  return run;
}

/**
 * Resolves coordinates to a readable address. Returns `undefined` when the
 * lookup fails; the caller is expected to fall back to the raw coordinates.
 */
export async function reverseGeocode(
  lat: number,
  lon: number,
  options: GeocodeOptions = {}
): Promise<string | undefined> {
  const key = entryKeyFor(lat, lon, options.places !== false, options.placeAddress === true);
  const hit = cache().get(key);
  if (hit) return hit.label;

  const server = await viaServer(lat, lon, options);
  // Switched off instance-wide. Returning here rather than falling through is
  // the point of the setting: the queue below would otherwise ask Nominatim
  // from the browser and undo the decision.
  if (server?.disabled) return undefined;
  if (server !== null) {
    // The server has its own cache; keeping a copy here saves the round trip
    // for the stays already on screen. An unsettled answer is kept only for
    // this page: the integration deliberately did not store it, and writing it
    // to browser storage for a month would undo that -- a shopping centre
    // would go on reading as the street outside it long after the lookup
    // service was willing again.
    if (server.label) {
      cache().set(key, {
        label: server.label,
        at: Date.now(),
        direct: !server.settled,
      });
    }
    persist();
    return server.label;
  }

  return schedule(async () => {
    // A second lookup may have filled the cache while this one was queued.
    const queued = cache().get(key);
    if (queued) return queued.label;

    const params = new URLSearchParams({
      format: "jsonv2",
      lat: String(lat),
      lon: String(lon),
      zoom: "18",
      addressdetails: "1",
    });
    if (options.language) params.set("accept-language", options.language);
    if (options.email) params.set("email", options.email);

    try {
      const response = await fetch(`${ENDPOINT}?${params.toString()}`, {
        headers: { Accept: "application/json" },
      });
      if (!response.ok) return undefined;
      const label = shortLabel(await response.json());
      if (!label) return undefined;
      cache().set(key, { label, at: Date.now(), direct: true });
      persist();
      return label;
    } catch {
      return undefined;
    }
  });
}

/** Test/debug helper: forgets every cached address. */
export function clearGeocodeCache(): void {
  memoryCache = undefined;
  // A write may already be queued, and it would put the cleared content
  // straight back. Dropping the handle also lets the next lookup queue one of
  // its own -- without this, the first write of a session was the last.
  if (persistHandle !== undefined) {
    try {
      window.clearTimeout(persistHandle);
    } catch {
      // A browser that cannot cancel it will still write nothing worth keeping.
    }
    persistHandle = undefined;
  }
  try {
    window.localStorage.removeItem(CACHE_KEY);
  } catch {
    // ignored
  }
}

File: custom_components/family_tracking/geocode.py
"""
Reverse geocoding, once for the whole household.

The card used to ask Nominatim straight from the browser. That works, but every
browser keeps its own cache, so the same street gets looked up again on the
phone, on the tablet and after every cleared browser storage -- and each of
those is a request against a service that asks for one per second, worldwide,
from everyone.

Doing it here means one cache for the whole instance, kept on disk across
restarts, and one queue that honours the rate limit no matter how many browsers
are open.

The cache lives in the integration's own database, next to the positions, and
is kept as long as they are: a month from half a year ago should come back with
its street names rather than a minute of queued lookups. It used to be a Home
Assistant `Store`, which is read once on the first start and then removed.
"""

from __future__ import annotations

import asyncio
import logging
import time
from typing import TYPE_CHECKING, Any

import aiohttp
from aiohttp import ClientError, ClientSession
from homeassistant.core import HomeAssistant
from homeassistant.helpers.storage import Store

from .address import Address, cache_key, merge_venue, parse
from .const import (
    CACHE_SCHEMA,
    MIN_REQUEST_INTERVAL,
    NOMINATIM_URL,
    OVERPASS_URLS,
    STORAGE_KEY,
    STORAGE_VERSION,
    VENUE_ATTEMPTS,
    VENUE_BACKOFF,
    VENUE_BUSY_STATUS,
    VENUE_CONNECT_TIMEOUT,
    VENUE_DEAD_FOR,
    VENUE_LIMIT_STATUS,
    VENUE_MIN_REQUEST_INTERVAL,
    VENUE_PAUSE,
    VENUE_TIMEOUT,
)
from .venue import build_query, covers, pick_name

if TYPE_CHECKING:
    from .track_store import TrackStore

_LOGGER = logging.getLogger(__name__)

# Re-exported so the rest of the integration keeps importing from one place.
__all__ = ["Address", "Geocoder", "cache_key", "merge_venue", "parse"]


class Geocoder:
    """One queue, one cache, one place that talks to Nominatim."""

    def __init__(
        self,
        hass: HomeAssistant,
        session: ClientSession,
        email: str | None,
        store: TrackStore,
    ) -> None:
        self._hass = hass
        self._session = session
        self._email = email
        self._store = store
        #: The file the cache lived in before it moved into the database.
        self._legacy: Store[dict[str, Any]] = Store(hass, STORAGE_VERSION, STORAGE_KEY)
        self._cache: dict[str, dict[str, Any]] = {}
        #: Keys changed since the last write; only these go to the database.
        self._changed: set[str] = set()
        # Serialises the requests. Without it a dashboard opening with twelve
        # stays would fire twelve lookups in the same tick.
        self._lock = asyncio.Lock()
        self._last_request = 0.0
        self._last_venue = 0.0
        #: Monotonic time before which Overpass is not to be asked again.
        self._venue_quiet_until = 0.0
        #: Which instance answered last; the next lookup starts there.
        self._venue_first = 0
        #: Instances that could not be reached, and until when to skip them.
        self._venue_dead: dict[str, float] = {}
        self._save_handle: asyncio.TimerHandle | None = None

    async def async_load(self) -> None:
        """Read the cache from the database, taking over the old file once."""
        self._cache = {
            key: entry
            for key, entry in (await self._store.async_load_addresses()).items()
            if isinstance(entry, dict) and entry.get("schema") == CACHE_SCHEMA
        }
        legacy = await self._legacy.async_load()
        if legacy:
            ttl = self._store.address_ttl()
            now = time.time()
            taken = {
                key: entry
                for key, entry in (legacy.get("entries") or {}).items()
                if isinstance(entry, dict)
                and entry.get("schema") == CACHE_SCHEMA
                and now - entry.get("at", 0) < ttl
                and key not in self._cache
            }
            await self._store.async_save_addresses(taken)
            self._cache.update(taken)
            await self._legacy.async_remove()
            _LOGGER.info("Moved %d cached address(es) into %s", len(taken), self._store.path)
        _LOGGER.debug("Loaded %d cached addresses", len(self._cache))

    async def async_save(self) -> None:
        if not self._changed:
            return
        changed, self._changed = self._changed, set()
        await self._store.async_save_addresses(
            {key: self._cache[key] for key in changed if key in self._cache}
        )

    def cached(self, latitude: float, longitude: float) -> Address | None:
        entry = self._cache.get(cache_key(latitude, longitude))
        if not entry:
            return None
        return merge_venue(Address(**entry["address"]), entry.get("venue") or "")

    async def async_resolve(
        self,
        latitude: float,
        longitude: float,
        language: str | None = None,
        places: bool = True,
        with_address: bool = False,
    ) -> Address | None:
        """
        What to call a position, from the cache where possible.

        The address and the name of the enclosing place are stored side by side
        rather than as one finished line. They answer different questions, the
        caller decides which it wants, and a dashboard that asks for one must
        not leave the other permanently unavailable to the next caller.
        """
        key = cache_key(latitude, longitude)
        entry = self._cache.get(key)
        if entry is not None and (not places or "venue" in entry):
            return self._compose(entry, places, with_address)

        async with self._lock:
            # Someone else may have filled it in while this call queued.
            entry = self._cache.get(key)
            if entry is not None and (not places or "venue" in entry):
                return self._compose(entry, places, with_address)

            wanted = []
            if entry is None:
                wanted.append(self._nominatim(latitude, longitude, language))
            if places:
                wanted.append(self._venue(latitude, longitude))
            # Both go to different services, so asking them at the same time
            # costs the longer of the two waits rather than their sum.
            answers = list(await asyncio.gather(*wanted))

        address = Address(**entry["address"]) if entry is not None else answers.pop(0)
        venue = answers.pop(0) if places else (entry or {}).get("venue")

        if address is None and not venue:
            return None

        stored: dict[str, Any] = {
            "address": (address or Address(label="")).as_dict(),
            "at": time.time(),
            "schema": CACHE_SCHEMA,
        }
        # `None` means the question could not be put, which is not the same as
        # "nothing is there". Storing it would freeze the street address in
        # place for months over one busy moment at the other end.
        if venue is not None:
            stored["venue"] = venue
        elif entry is not None and "venue" in entry:
            stored["venue"] = entry["venue"]

        if address is not None:
            self._cache[key] = stored
            self._changed.add(key)
            self._schedule_save()

        endgueltig = merge_venue(address, venue or "", with_address)
        if endgueltig is not None and places and venue is None:
            # Overpass could not be asked. The address goes back as a usable
            # line, but marked so that nothing downstream stores it -- the
            # browser keeps its own copy for a month, which would outlive the
            # busy moment by a long way.
            endgueltig.settled = False
        return endgueltig

    @staticmethod
    def _compose(entry: dict[str, Any], places: bool, with_address: bool = False) -> Address:
        address = Address(**entry["address"])
        if not places:
            return address
        return merge_venue(address, entry.get("venue") or "", with_address)

    async def _nominatim(
        self, latitude: float, longitude: float, language: str | None
    ) -> Address | None:
        # The rate limit lives here rather than around the call, because the two
        # services are now asked at the same time and each has its own pace to
        # keep. Nominatim's usage policy is one request per second.
        wait = self._last_request + MIN_REQUEST_INTERVAL - time.monotonic()
        if wait > 0:
            await asyncio.sleep(wait)
        self._last_request = time.monotonic()

        params = {
            "format": "jsonv2",
            "lat": f"{latitude}",
            "lon": f"{longitude}",
            "zoom": "18",
            "addressdetails": "1",
        }
        if language:
            params["accept-language"] = language
        if self._email:
            params["email"] = self._email

        try:
            async with self._session.get(
                NOMINATIM_URL,
                params=params,
                headers={"User-Agent": "home-assistant-family-tracking"},
                timeout=15,
            ) as response:
                if response.status != 200:
                    _LOGGER.debug("Nominatim answered %s", response.status)
                    return None
                payload = await response.json(content_type=None)
        except (ClientError, asyncio.TimeoutError) as err:
            _LOGGER.debug("Nominatim unreachable: %s", err)
            return None

        address = parse(payload)
        return address if address.label else None

    async def _venue(self, latitude: float, longitude: float) -> str | None:
        """
        The name of the place this coordinate is inside.

        Three outcomes, and the difference between the last two matters: a name,
        `""` for "nothing encloses this spot", and `None` for "could not ask".
        Overpass is donated capacity and answers a burst with 429; treating that
        like an empty result would write the street address into a cache that
        holds for months, and the shopping centre would stay misnamed long after
        the service was happy again.
        """
        if time.monotonic() < self._venue_quiet_until:
            # Told to slow down recently. Asking anyway is how a busy signal
            # turns into a block.
            return None

        query = build_query(latitude, longitude)

        # Each instance gets the full round of attempts before the next one is
        # tried. A rate limit ends that instance's turn immediately -- asking
        # again is the one thing it just told us not to do.
        reihenfolge = [
            OVERPASS_URLS[(self._venue_first + i) % len(OVERPASS_URLS)]
            for i in range(len(OVERPASS_URLS))
        ]

        for stelle, url in enumerate(reihenfolge):
            if time.monotonic() < self._venue_dead.get(url, 0.0):
                continue
            gedrosselt = False
            for versuch in range(1, VENUE_ATTEMPTS + 1):
                wait = self._last_venue + VENUE_MIN_REQUEST_INTERVAL - time.monotonic()
                if wait > 0:
                    await asyncio.sleep(wait)
                self._last_venue = time.monotonic()

                try:
                    async with self._session.post(
                        url,
                        data={"data": query},
                        headers={"User-Agent": "home-assistant-family-tracking"},
                        timeout=aiohttp.ClientTimeout(
                            total=VENUE_TIMEOUT, sock_connect=VENUE_CONNECT_TIMEOUT
                        ),
                    ) as response:
                        if response.status in VENUE_LIMIT_STATUS:
                            _LOGGER.debug("%s is rate limiting (%s)", url, response.status)
                            gedrosselt = True
                            break
                        if response.status in VENUE_BUSY_STATUS:
                            _LOGGER.debug(
                                "%s is busy (%s), attempt %d of %d",
                                url, response.status, versuch, VENUE_ATTEMPTS,
                            )
                            continue
                        if response.status != 200:
                            _LOGGER.debug("%s answered %s", url, response.status)
                            break
                        # Overpass reports its own errors as XML with a 200, so
                        # the content type is not something to insist on here
                        # -- but then the body will not parse, which lands in
                        # the same place as any other failure.
                        payload = await response.json(content_type=None)
                except (ClientError, asyncio.TimeoutError, ValueError) as err:
                    _LOGGER.debug("Could not reach %s: %s", url, err)
                    self._venue_dead[url] = time.monotonic() + VENUE_DEAD_FOR
                    break

                if not covers(payload):
                    # It answered, but it has no data for this part of the
                    # world. Believing it would cache "nothing here" as fact.
                    _LOGGER.debug("%s holds no data around this fix", url)
                    break

                # Worth starting here next time.
                self._venue_first = (self._venue_first + stelle) % len(OVERPASS_URLS)
                return pick_name(payload)

            if gedrosselt and stelle == len(reihenfolge) - 1:
                # Every instance turned us away. That is about us, not them.
                self._venue_quiet_until = time.monotonic() + VENUE_BACKOFF
                return None

        # Nobody answered usefully. A short pause, not the long one.
        self._venue_quiet_until = time.monotonic() + VENUE_PAUSE
        return None

    def _schedule_save(self) -> None:
        """Write at most once every half minute rather than per lookup."""
        if self._save_handle is not None:
            return

        def _write() -> None:
            self._save_handle = None
            self._hass.async_create_task(self.async_save())

        self._save_handle = self._hass.loop.call_later(30, _write)
